src/puppetlabs/puppetdb/query/facts.clj

```diff
@@ -31,26 +28,21 @@
 
 ;; MUNGE
 
-(pls/defn-validated convert-types :- [converted-row-schema]
+(pls/defn-validated deserialize-fact-value
   "Coerce values for each row to the proper stored type."
-  [rows :- [row-schema]]
-  (map (partial facts/convert-row-type [:type :depth :value_integer :value_float]) rows))
+  [row]
+  (utils/update-when row [:value] json/parse-string))
 
-(pls/defn-validated munge-result-rows
-  [_
-   projected-fields :- [s/Keyword]
-   _
-   _]
+(defn munge-result-rows
+  [_ _ _]
   (fn [rows]
     (if (empty? rows)
       []
       (->> rows
-        convert-types
-        (map #(select-keys % (or (seq projected-fields)
-                                 [:certname :environment :timestamp :name :value])))))))
+           (map deserialize-fact-value)))))
 
 (defn munge-path-result-rows
-  [_ _ _ _]
+  [_ _ _]
```

Comment:
the need isn't currently apparent, but it'll come back as soon as we cut more API versions after the 3.0 release. The main reason it looks strange now is because we recently dropped v2 and v3, which previously required special casing. I'm a sort of ambivalent about optimizing for our current situation for this reason, although I definitely see where you're coming from.